rasa/core/policies/ted_policy.py

```diff
@@ -1092,36 +1095,84 @@ def _compute_dialogue_indices(
             ).values
         ]
 
-    def _create_all_labels_embed(self) -> Tuple[tf.Tensor, tf.Tensor]:
+    def _create_all_labels_embed_for_training(
+        self, number_of_random_labels: int
+    ) -> Tuple[tf.Tensor, tf.Tensor]:
+
         all_label_ids = self.tf_label_data[LABEL_KEY][LABEL_SUB_KEY][0]
-        # labels cannot have all features "fake"
-        all_labels_encoded = {}
+
+        if number_of_random_labels != -1:
+
+            # First, sample a fixed number of random labels ids
+            total_num_label_ids = all_label_ids.shape[0]
+            sampled_label_ids, _, _ = tf.random.uniform_candidate_sampler(
+                tf.reshape(
+                    tf.cast(tf.range(0, total_num_label_ids), dtype=tf.int64),
+                    shape=(1, -1),
+                ),
+                total_num_label_ids,
+                number_of_random_labels,
+                True,
+                total_num_label_ids,
+            )
+            sampled_label_ids = tf.reshape(sampled_label_ids, (-1, 1))
+
+        else:
+            sampled_label_ids = all_label_ids
+
+        # Compute embeddings for sampled labels
+        sampled_labels_embed = self._compute_embedding_for_label_ids(sampled_label_ids)
+
+        return sampled_label_ids, sampled_labels_embed
+
+    def _slice_tensor(
+        self, tensor: Union[tf.Tensor, tf.SparseTensor], indices: tf.Tensor
+    ) -> Union[tf.Tensor, tf.SparseTensor]:
+        if isinstance(tensor, tf.Tensor):
+            return tf.gather_nd(tensor, tf.cast(indices, dtype=tf.int32))
+        elif isinstance(tensor, tf.SparseTensor):
+            dense_tensor = tf.sparse.to_dense(tensor)
+            sliced_tensor = tf.gather_nd(dense_tensor, tf.cast(indices, dtype=tf.int32))
+            return tf.sparse.from_dense(sliced_tensor)
```

Comment:
This will blow up the memory a bit but not sure what else can be done because `tf.gather_nd` does not support sparse tensors.